### DISCUSSION TITLE
Sourcing schema for all-teh-things?

### DISCUSSION THREAD

**Discussion by rgeyer** [pollypkg/polly#34] [Q&A] [answered]
I'm stumbling down the trail(s) you've blazed WRT to grafana dashboard schema/cue/polly, and would like to validate a few assertions.

Specifically, I am trying to formalize some "standards" for Grafana Cloud integrations. Specifically around available, and preferred visualizations/panels.

So..

1. The entirety of the grafana schema lives in [grafana/grafana/cue/data/gen.cue](https://github.com/grafana/grafana/blob/main/cue/data/gen.cue)?
2. That file suggests that's it's generated? How? Where?
3. That file only contains a schema for a handful of panel/visualization types, Panel, GraphPanel, and RowPanel. It occurs to me that this is "by design" since all other visualization types are actually plugins. Are there schemas for those panel types?

I'm trying my best not to have to hand-curate a list of available, and preferred visualization types, to compare the existing mixins/dashboards against, but... It **_feels_** like I'm going to have to :thinking:

**Comment by rgeyer** [comment]
RE: 3, I found that the schema(s) do exist in public/app/plugins/panel/**/models.cue. Like [this one](https://github.com/grafana/grafana/blob/main/public/app/plugins/panel/barchart/models.cue) for barchart.

Same question remains for 2, related to each of those panel types. :thinking:

**Reply by rgeyer** [reply]
Q&A is probably the wrong type of discussion for this, now that I'm adding more comments. At any rate....

It seems that the schema as-is won't solve my problem. The schema only adds rules which define the constraints for particular things. So for instance, when I add the barchart cue file, it only adds constrains that are specific to the barchart panel.

Also, the panel schema definition itself only defines that the panel plugin type [must be defined](https://github.com/grafana/grafana/blob/main/cue/data/gen.cue#L139), but not which values are valid for that property.

This all seems reasonable, and by design I suppose.

Is it the intent that polly could produce a dashboard/alert/rule artifact which wouldn't pass validation when actually submitted to grafana/prometheus/cortex?

**Reply by sdboyer** [reply]
> Is it the intent that polly could produce a dashboard/alert/rule artifact which wouldn't pass validation when actually submitted to grafana/prometheus/cortex?

No, having this be impossible is pretty much the most foundational guarantee we're aiming for. 

(other q's here should be answered by my big reply below)

**Comment by sdboyer** [comment, marked as answer]
Yeah, there's a bunch of stuff that's half done. In general i haven't been doing a good job in making these temporary states clear by way of issues posted in polly and/or grafana, though i've been turning that around (at least with grafana) in the last week. May the trend continue :)

To each item:

> The entirety of the grafana schema lives in grafana/grafana/cue/data/gen.cue?

This - or at least, somewhere under `cue/` - is the publicly importable place that the scuemata will live. The canonical version that's directly managed and updated by Grafana core devs will live somewhere else - more details in grafana/grafana#38033 - but for our purposes in polly, yeah, this is where the whole scuemata schema history will live.

> That file suggests that's it's generated? How? Where?

See the above issue. It's the intent that it be generated, but we've been sloppy in this first phase :)

> That file only contains a schema for a handful of panel/visualization types, Panel, GraphPanel, and RowPanel. It occurs to me that this is "by design" since all other visualization types are actually plugins. Are there schemas for those panel types?

Most complicated part, by far! i alluded to this in #31, but i'll expand specifically, here.

The core dashboard schema - [this](https://github.com/grafana/grafana/blob/main/cue/data/gen.cue) - is defined without any explicit reference to the schema declared by plugins. This is a design necessity, because the possibility of third party plugins makes it impossible to statically enumerate, import, and compose in all possible plugins. Instead, we do that dynamically, at runtime.

[This gist](https://gist.github.com/sdboyer/a8a5eacb9ceac45aca5c77cb7a34661b) is a dump of what that dynamically (in Go code) composed dashboard schema looks like, having pulled in all the schema currently defined in core plugins. (Note that a bunch of things are absent, because the default printing behavior omits optional fields, and we haven't yet written a custom one. That's a TODO, and once it's done, we'll add a `grafana-cli` command that dumps this out properly.)

The dynamic schema contains a huge disjunction ([lines 62-409](https://gist.github.com/sdboyer/a8a5eacb9ceac45aca5c77cb7a34661b#file-dashboard-cue-L62-L409)) which collectively comprises all the versions of all the core plugins. Look through, and you'll see each one, identified primarily by the `type` field - e.g. here's [barchart](https://gist.github.com/sdboyer/a8a5eacb9ceac45aca5c77cb7a34661b#file-dashboard-cue-L63), [histogram](https://gist.github.com/sdboyer/a8a5eacb9ceac45aca5c77cb7a34661b#file-dashboard-cue-L214), etc.

The above is the normal, expected pattern. The dashboard schema directly contains `#RowPanel` and `#GraphPanel` as one-off hacks, because they're abnormal and don't follow the same compositional rules as we're henceforth enforcing that all panel plugins must follow.

The challenge in #31 is about replicating the Go logic that does this dynamic composition process, in CUE, so that we can produce the same logical schema. Eventually, i hope this kind of composition will be a standard pattern structured by the rules of scuemata itself, but we can hack it in for now.

But you do highlight something important, here:

> I'm trying my best not to have to hand-curate a list of available, and preferred visualization types, to compare the existing mixins/dashboards against, but... It feels like I'm going to have to 🤔

This is an important challenge - the non-enumerability of all plugins will affect Polly in the same way it affects Grafana core: we can't possibly write the Polly schema in a way that includes all third party plugins, wherever they may be defined. The only real choice we have here is to leave the schema open in such a way that, if the `type` field of a panel plugin (and it'll be the same for datasource plugins) isn't one of the values known to the schema, then the parts normally specified by a plugin will just be left open, allowing anything to be there.

There are some steps we can take to tighten that up, but it's further down the road. For a while, the sanest path is going to be to only use core Grafana plugins when writing a pop. That's true even if the core plugin isn't currently in that gist, e.g. [alertlist](https://github.com/grafana/grafana/tree/main/public/app/plugins/panel/alertlist) - it's pretty trivial to add a `models.cue` to core plugins, and we'll be doing that rapidly in the coming weeks. So, at least, no hand curation necessary - just limit yourself to core plugins, and if you need one that's isn't in that gist, file an issue against Grafana asking for one to be added :)
